internal/osbuild2/tmpfilesd_stage.go

```diff
@@ -0,0 +1,61 @@
+package osbuild2
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// TmpfilesdStageOptions represents a single tmpfiles.d configuration file.
+type TmpfilesdStageOptions struct {
+	// Filename of the configuration file to be created. Must end with '.conf'.
+	Filename string `json:"filename"`
+	// List of configuration directives. The list must contain at least one item.
+	Config []TmpfilesdConfigLine `json:"config"`
+}
+
+func (TmpfilesdStageOptions) isStageOptions() {}
+
+// NewTmpfilesdStageOptions creates a new Tmpfilesd Stage options object.
+func NewTmpfilesdStageOptions(filename string, config []TmpfilesdConfigLine) *TmpfilesdStageOptions {
+	return &TmpfilesdStageOptions{
+		Filename: filename,
+		Config:   config,
+	}
+}
+
+// Unexported alias for use in TmpfilesdStageOptions's MarshalJSON() to prevent recursion
+type tmpfilesdStageOptions TmpfilesdStageOptions
+
+func (o TmpfilesdStageOptions) MarshalJSON() ([]byte, error) {
+	if len(o.Config) == 0 {
+		return nil, fmt.Errorf("the 'Config' list must contain at least one item")
+	}
+	options := tmpfilesdStageOptions(o)
+	return json.Marshal(options)
+}
```

Comment:
Thanks for pointing this out, I didn't notice your approach 🙄 . I agree with everything that you've wrote.

In addition, my motivation of handling this in `MarshalJSON()` was that this is the last step in the code path when we are creating an osbuild manifest, regardless of the way the stage structure has been created (using the `NewXStage()` or "manually"). We should use `NewXStage()` everywhere in the code, but it is not used at least in `internal/osbuild2/stage_test.go` unit tests, where the stages are created as structure literals. By validating the data in `MarshalJSON()`, I basically made sure that I could not test invalid data in `stage_test.go`. Probably not a huge benefit, but I just wanted to mention it.

We could also have something in-between your and my approach - have a separate `validate()` method and just call it in the `MarshalJSON()`. Not that we are going to marshal to anything else than JSON in the near future, but this would make the validation code probably clearer and more reusable if needed.

Reply:
I think I prefer the validation in `NewXStage` method because then it's not possible to create a stage with invalid options. I say: Fail fast!

Problems (and my proposed solutions):

- `NewXStage()` method panics: Either we can just live with that (usually it's just a programming error so it's fine to panic), or we should just return also `error` from this method.
- People can always use `XStageOptions{}` directly instead of using `NewXStage()`: Make `XStageOptions` private: `xStageOptions`. Some adjustments need to be done to the code but I think it might be worth it since the code is imho safer.

Reply:
I'm OK with your proposal @ondrejbudai, however I'd prefer to do it as a follow-up. If this is to be done for all stages, it will be quite a lot of work. I would like to have the 9.0 EC2 SAP image defined by the end of this month, so that EXD can ideally use it for Beta images, since they are still blocked on the availability of RHUI client rpms.

Reply:
Agreed.  Opening issue to unify validation.